metis-remove-data/src/main/java/eu/europeana/metis/remove/discover/ExecutionPluginForest.java

```diff
@@ -57,39 +63,71 @@
+        (ignoreNode ? ignoredNodesById : nodesById).put(node.getId(), node);
       }
     }
 
-    // Add all nodes to the lookup map for predecessors
-    final Function<ExecutionPluginNode, PluginLinkByTypeAndTimestamp> extractTimeAndTimestamp = node -> new PluginLinkByTypeAndTimestamp(
-        node.getType(), node.getPlugin().getStartedDate());
-    this.nodesByTypeAndTimestamp = nodesById.values().stream()
-        .filter(node -> node.getPlugin().getStartedDate() != null)
-        .collect(Collectors.toMap(extractTimeAndTimestamp, Function.identity()));
-
-    // Link parents and children (predecessors and successors)
-    for (ExecutionPluginNode node : nodesById.values()) {
-      final ExecutionPluginNode predecessor = findPredecessor(node);
-      if (predecessor != null) {
-        predecessor.addChild(node);
+    // Add all nodes to the lookup map for predecessors (also ignored ones: they could still have
+    // children that are not to be ignored).
+    final Function<ExecutionPluginNode, PluginLinkByTypeAndTimestamp> extractTimeAndTimestamp =
+            node -> new PluginLinkByTypeAndTimestamp(node.getType(),
+                    node.getPlugin().getStartedDate());
+    final Map<PluginLinkByTypeAndTimestamp, ExecutionPluginNode> predecessorLookup = Stream
+            .concat(nodesById.values().stream(), ignoredNodesById.values().stream())
+            .filter(node -> node.getPlugin().getStartedDate() != null)
+            .collect(Collectors.toMap(extractTimeAndTimestamp, Function.identity()));
+
+    // Link parents and their children (predecessors and successors). This happens in iterations
+    // because during linking we may find nodes that we planned to ignore but still have children
+    // that we don't want to ignore. We loop until we find no such nodes anymore.
+    final Function<String, ExecutionPluginNode> nodeByIdResolver = id -> Optional
+            .ofNullable(nodesById.get(id)).orElseGet(() -> ignoredNodesById.get(id));
+    Map<String, ExecutionPluginNode> nodesToProcess = nodesById;
+    while (!nodesToProcess.isEmpty()) {
+
+      // We will need a list of nodes that we planned to ignore but need to keep after all.
+      final Map<String, ExecutionPluginNode> ignoredNodesToBeKept = new HashMap<>();
+
+      // We go by all nodes in this iteration and attempt to find the parent.
+      for (ExecutionPluginNode node : nodesToProcess.values()) {
+        final ExecutionPluginNode predecessor = findPredecessor(node, nodeByIdResolver,
+                predecessorLookup::get);
+        if (predecessor != null) {
+
+          // So we have a parent. If we don't have it in our node map yet, we mark it for
+          // processing in the next iteration. We don't add it directly to the node map as we may be
+          // iterating over that very collection (during the first iteration).
+          if (!nodesById.containsKey(predecessor.getId())) {
```

Comment:
Can this actually happen? That a predecessor plugin status is DELETED and it's children are not?

Reply:
Not currently, but there is also nothing to prevent us from doing this (on a manual/incidental basis). So I wanted to make sure that we can support it, just in case.